### PR TITLE
[BE-05] Add comments

### PR DESCRIPTION
**Pull request by Enriquee98**
https://trello.com/c/SfakkMcW

### CODE REVIEW

File: app/controllers/api/v1/recipes_controller.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@ def index
       end
 
       def show
+        @recipe.cooking_time ||= 'unknow'
```
Comment (grabomarcin) [inline app/controllers/api/v1/recipes_controller.rb]: 👀

Comment (plokijro) [inline app/controllers/api/v1/recipes_controller.rb]: What's going on here? 😁

File: app/serializers/comment_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class CommentSerializer < ActiveModel::Serializer
+  attributes :id, :author, :body, :created_at
+ ANON = 'annonymous'
```
Comment (grabomarcin) [inline app/serializers/comment_serializer.rb]: Use some better name for that const and move that above `attributes` :p

File: app/serializers/recipe_serializer_show.rb [hunk 3]
```diff
@@ -0,0 +1,3 @@
+class RecipeSerializerShow < ActiveModel::Serializer
+  attributes :id, :name, :content, :cooking_time
+end
```
Comment (grabomarcin) [inline app/serializers/recipe_serializer_show.rb]: You can inherit that serializer from `RecipeSerializer` and just add what is missing.

File: skrypt.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+
+IP_ADDR=`/sbin/ifconfig eth0 | grep 'inet addr' | cut -d: -f2 | awk '{print $1}'`
+
+echo "Twoj publiczny adres ip to: "
+curl http://169.254.169.254/latest/meta-data/public-ipv4 
+echo ""
+rails s -b $IP_ADDR -p 8082
```
Comment (grabomarcin) [inline skrypt.sh]:
```suggestion
rails s -b $IP_ADDR -p 8082

```

File: app/serializers/recipe_serializer.rb [hunk 5]
```diff
@@ -1,3 +1,8 @@
 class RecipeSerializer < ActiveModel::Serializer
   attributes :id, :name, :content
+  has_many :comments
+
+  def comments
+    object.comments.order(created_at: :desc)
+  end
```
Comment (plokijro) [inline app/serializers/recipe_serializer.rb]: Comments should be returned only in show recipe endpoint

File: spec/controllers/api/v1/comment_serializer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    context 'when comment with author is present' do
+      it {is_expected.to include(author: comment.author)}
+    end
+    
+    context 'when comment with author is nil' do
+      let(:comment) {build :comment, author: nil }
+      it {is_expected.to include(author: 'annonymous')}
+    end
+  end
+end
```
Comment (plokijro) [inline spec/controllers/api/v1/comment_serializer_spec.rb]: Run rubocop please

Comment (grabomarcin) [inline spec/controllers/api/v1/comment_serializer_spec.rb]:
```suggestion
require 'rails_helper'

RSpec.describe CommentSerializer do
  subject { described_class.new(comment).to_h }
  
  let(:comment) { build :comment }
  
  it { is_expected.to include(body: comment.body) }
  it { is_expected.to include(id: comment.id) }
  it { is_expected.to include(created_at: comment.created_at) }
  
  describe '#author' do
    context 'when comment with author is present' do
      it { is_expected.to include(author: comment.author) }
    end
    
    context 'when comment with author is nil' do
      let(:comment) { build :comment, author: nil }
      it { is_expected.to include(author: described_class::CONST_WHICH_YOU_DEFINED) }
    end
  end
end

```

You can call const here. Please use rubocop to fix syntax

File: spec/factories/comments.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :comment do
+    author {Faker::Name.name}
+    body {Faker::ChuckNorris.fact}
+    recipe
+  end
+end
```
Comment (grabomarcin) [inline spec/factories/comments.rb]:
```suggestion
end

```

File: spec/models/comment_spec.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+require 'rails_helper'
+
+RSpec.describe Comment, type: :model do
+  describe 'association' do
```
Comment (grabomarcin) [inline spec/models/comment_spec.rb]:
```suggestion
  describe 'associations' do
```

File: app/models/recipe.rb [hunk 9]
```diff
@@ -1,3 +1,5 @@
 class Recipe < ApplicationRecord
+  has_many :comments, dependent: :destroy
   validates :name, presence: true
+  validates :cooking_time, numericality: { greater_than: 0 }
 end
```
Comment (plokijro) [inline app/models/recipe.rb]: `cooking_time`? Isn't it from another task?

File: app/serializers/comment_serializer.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+class CommentSerializer < ActiveModel::Serializer
+  ANONYMOUS = 'annonymous'.freeze
+  attributes :id, :author, :body, :created_at
+  def author
+    return ANONYMOUS unless object.author
+
+    object.author
+  end
```
Comment (plokijro) [inline app/serializers/comment_serializer.rb]:
```suggestion
  attributes :id, :author, :body, :created_at
  
  def author
    return ANONYMOUS unless object.author

    object.author
  end
```

File: app/serializers/recipe_show_serializer.rb [hunk 11]
```diff
@@ -0,0 +1,15 @@
+class RecipeShowSerializer < RecipeSerializer
+  NIL_RESULT = 'unknown'.freeze
+  attributes :cooking_time
+  has_many :comments
```
Comment (plokijro) [inline app/serializers/recipe_show_serializer.rb]:
```suggestion
  attributes :cooking_time
  
  has_many :comments
```

File: spec/factories/comments.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+FactoryBot.define do
+  factory :comment do
+    author { Faker::Name.name }
+    body { Faker::ChuckNorris.fact }
+    recipe
```
Comment (plokijro) [inline spec/factories/comments.rb]:
```suggestion
    recipe
    author { Faker::Name.name }
    body { Faker::ChuckNorris.fact }
```
Associations should be before attributes

File: app/controllers/api/v1/comments_controller.rb [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+      before_action :set_recipe, only: %i[create]
+
+      def create
+        comment = Comment.new(comment_params)
+        comment.recipe = @recipe
+
+        if comment.save
+          render json: comment
+        else
+          render json: { errors: comment.errors.to_s }, status: :unprocessable_entity
```
Comment (grabomarcin) [inline app/controllers/api/v1/comments_controller.rb]:
```suggestion
          render json: { errors: comment.errors.full_messages }, status: :unprocessable_entity
```
IMO it's better

File: app/controllers/api/v1/recipes_controller.rb [hunk 14]
```diff
@@ -38,7 +38,9 @@ def destroy
       private
 
       def recipe_params
-        params.require(:recipe).permit(%i[name content video_link])
+
```
Comment (grabomarcin) [inline app/controllers/api/v1/recipes_controller.rb]:
```suggestion
```

File: app/models/recipe.rb [hunk 15]
```diff
@@ -1,4 +1,8 @@
 class Recipe < ApplicationRecord
+  has_many :comments, dependent: :destroy
   validates :name, presence: true
```
Comment (grabomarcin) [inline app/models/recipe.rb]:
```suggestion

  validates :name, presence: true
```

File: app/models/recipe.rb [hunk 16]
```diff
@@ -1,4 +1,8 @@
...
-  validates :video_link, allow_blank: true, url: true
+
+  validates :cooking_time, numericality: { greater_than: 0 }
+   validates :video_link, allow_blank: true, url: true
+
```
Comment (grabomarcin) [inline app/models/recipe.rb]:
```suggestion
  validates :cooking_time, numericality: { greater_than: 0 }
  validates :video_link, allow_blank: true, url: true
```

File: app/serializers/comment_serializer.rb (hunk 10 above)
Comment (grabomarcin) [inline app/serializers/comment_serializer.rb]:
```suggestion
  ANONYMOUS = 'annonymous'.freeze
  
  attributes :id, :author, :body, :created_at
  
  def author
    return ANONYMOUS unless object.author

    object.author
  end
```

File: app/serializers/recipe_show_serializer.rb [hunk 17]
```diff
@@ -1,11 +1,22 @@
...
+
+  def comments
+    object.comments.order(created_at: :desc)
+    
   def video_link
     return NIL_RESULT unless object.video_link
 
     object.video_link
+
   end
```
Comment (grabomarcin) [inline app/serializers/recipe_show_serializer.rb]: 👀

File: db/migrate/20220713071428_create_comments.rb [hunk 18]
```diff
@@ -0,0 +1,12 @@
+class CreateComments < ActiveRecord::Migration[6.1]
+  def change
+    create_table :comments do |t|
+      t.string :author
+      t.string :body, null: false
+      t.belongs_to :recipe, foreign_key: true
+
```
Comment (grabomarcin) [inline db/migrate/20220713071428_create_comments.rb]:
```suggestion
```

File: spec/factories/comments.rb (hunk 12 above)
Comment (grabomarcin) [inline spec/factories/comments.rb]:
```suggestion
    recipe
    author { Faker::Name.name }
    body { Faker::ChuckNorris.fact }
```

**Review by grabomarcin — CHANGES_REQUESTED** [review]

File: app/models/comment.rb [hunk 19]
```diff
@@ -1,4 +1,5 @@
 class Comment < ApplicationRecord
   belongs_to :recipe
   validates :body, presence: true, length: { minimum: 10 }
+  validates :author, allow_blank: true, length: { minimum: 3 }
```
Comment (grabomarcin) [inline app/models/comment.rb]:
```suggestion
  belongs_to :recipe

  validates :body, presence: true, length: { minimum: 10 }
  validates :author, allow_blank: true, length: { minimum: 3 }
```

File: spec/serializers/comment_serializer_spec.rb [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+  describe '#comment' do
+    context ' with author' do
+      it { is_expected.to include(author: comment.author) }
+    end
+
+    context ' without author' do
+      let!(:comment) { create :comment, author: nil }
+
+      it { is_expected.to include(author: described_class::ANONYMOUS) }
+    end
```
Comment (plokijro) [inline spec/serializers/comment_serializer_spec.rb:21]:
```suggestion
    context 'with author' do
      it { is_expected.to include(author: comment.author) }
    end

    context 'without author' do
      let!(:comment) { create :comment, author: nil }

      it { is_expected.to include(author: described_class::ANONYMOUS) }
    end
```

File: app/models/recipe.rb [hunk 21]
```diff
@@ -1,9 +1,11 @@
...
 
   validates :name, presence: true
-  validates :level, allow_nil: true,
+  validates :cooking_time, numericality: { greater_than: 0 }, allow_nil: true
+  validates :video_link, allow_blank: true, url: true
+    validates :level, allow_nil: true,
                     inclusion: { in: 1..5,
                                  message: INCLUSION_ERROR_MESSAGE }
-  validates :video_link, allow_blank: true, url: true
+  INCLUSION_ERROR_MESSAGE = 'must be integer in range 1-5 inclusive'.freeze
```
Comment (grabomarcin) [inline app/models/recipe.rb]: Move const definition up

File: app/models/recipe.rb [hunk 22]
```diff
@@ -1,9 +1,11 @@
 class Recipe < ApplicationRecord
-  INCLUSION_ERROR_MESSAGE = 'must be integer in range 1-5 inclusive'.freeze
+  has_many :comments, dependent: :destroy
 
   validates :name, presence: true
-  validates :level, allow_nil: true,
+  validates :cooking_time, numericality: { greater_than: 0 }, allow_nil: true
+  validates :video_link, allow_blank: true, url: true
+    validates :level, allow_nil: true,
```
Comment (grabomarcin) [inline app/models/recipe.rb]: Fix syntax

File: app/serializers/recipe_show_serializer.rb [hunk 23]
```diff
@@ -1,7 +1,18 @@
 class RecipeShowSerializer < RecipeSerializer
   NIL_RESULT = 'unknown'.freeze
 
-  attributes :level, :video_link
+
```
Comment (grabomarcin) [inline app/serializers/recipe_show_serializer.rb]: unnecessary empty line

File: app/models/recipe.rb [hunk 24]
```diff
@@ -1,9 +1,11 @@
 class Recipe < ApplicationRecord
   INCLUSION_ERROR_MESSAGE = 'must be integer in range 1-5 inclusive'.freeze
```
Comment (grabomarcin) [inline app/models/recipe.rb:2]:
```suggestion
  INCLUSION_ERROR_MESSAGE = 'must be integer in range 1-5 inclusive'.freeze
  
```

File: app/serializers/recipe_show_serializer.rb [hunk 25]
```diff
@@ -1,11 +1,22 @@
...
+  end
+
+  def comments
+    object.comments.order(created_at: :desc)
+  end
 
   def level
     return NIL_RESULT unless object.level
-
+
```
Comment (grabomarcin) [inline app/serializers/recipe_show_serializer.rb]:
```suggestion

```

File: app/serializers/recipe_show_serializer.rb [hunk 26]
```diff
@@ -1,11 +1,22 @@
 class RecipeShowSerializer < RecipeSerializer
   NIL_RESULT = 'unknown'.freeze
 
-  attributes :level, :video_link
+  attributes :cooking_time, :video_link, :level
+  has_many :comments
```
Comment (grabomarcin) [inline app/serializers/recipe_show_serializer.rb]:
```suggestion
  attributes :cooking_time, :video_link, :level

  has_many :comments
```

**Review by grabomarcin — APPROVED** [review]
LGTM

**Pull request merged**